client.go

```diff
@@ -976,6 +985,8 @@ func clientDoDeadline(req *Request, resp *Response, deadline time.Time, c client
 		ReleaseRequest(reqCopy)
 		errorChPool.Put(chv)
 	case <-tc.C:
+		atomic.StoreInt32(&cleanup, 1)
+		errorChPool.Put(chv)
```

Comment:
I just figured out that this line can cause a lot of problems. What if the request had an error just before the timeout. In that case the request goroutine will write to `ch` but the timeout case will put `ch` back into the pool. So now we have a channel with a value already in there in the pool. This will then cause the next request to get the channel from the pool to immediately hit the case with the previous error.
The solution is to move this line to line `970` so we only reuse the channel if we know a cleanup is needed and we know we haven't written to it yet.

Reply:
but the second case will never be executed, 
only one case will be executed ??

Reply:
Why wouldn't the second case be executed? Since its two different goroutines its definitely possible that one is just entering the timeout case without having changed cleanup yet, and the other goroutine is just after the cleanup check but hasn't put the value into the channel yet.

Reply:
Ok, I've already moved the channel pool release to the upper coroutine as suggested